TL-3.1-HSBM.py

```diff
@@ -0,0 +1,291 @@
+        min_clusters=1,
+        max_clusters=10,
+        n_components=None,
+        n_elbows=2,
+        normalize=False,
+        regularizer=None,
+        min_split=10,
+        plus_c=False,
+    ):
+        super().__init__()
+        self.name = name
+        self.parent = parent
+        self.ifembed = ifembed
+        self.reembed = reembed
+        self.n_init = n_init
+        self.min_clusters = min_clusters
+        self.max_clusters = max_clusters
+        self.n_components = n_components
+        self.n_elbows = n_elbows
+        self.normalize = normalize
+        self.embed_method = embed_method
+        self.cluster_method = cluster_method
+        self.regularizer = regularizer
+        self.plus_c = plus_c
+        self.min_split = min_split
+        self.selection_criteria = selection_criteria
+        self.X = X
+        self.adj = adj
+
+        # if cluster_method not in ["GMM", "KMeans", "Spherical-Kmeans"]:
+        #     msg = "clustering method must be one of [GMM, Kmeans, Spherical-Kmeans]"
+        #     raise ValueError(msg)
+
+        # if embed_method not in ["ASE", "LSE"]:
+        #     msg = "embedding method must be either ASE or LSE"
+        #     raise ValueError(msg)
+
+        labels = range(len(self.X))
+        labels = pd.DataFrame(labels)
+        labels.columns = ["inds"]
+        self.labels = labels
+
+        if root_inds is None:
+            print("No `root_inds` were input; assuming this is the root.")
+            root_inds = labels["inds"]
+        self.root_inds = root_inds
+
+    def fit(self):
+        while True:
+            all_k = []
+            for i, node in enumerate(self.get_lowest_level()):
+                node.fit_candidates()
+                node.select_model()
+                all_k.append(node.k_)
+            self.collect_labels()
+            if all_k.count(1) == len(all_k):
+                break
+
+        mask = self.labels.columns.str.endswith(('pred'))
+        self.labels = self.labels.loc[:, ~mask]
+
+    def fit_candidates(self):
+        if hasattr(self, "k_") and self.k_ == 0:
+            return
+
+        if self.is_root:
+            if self.ifembed is True:
+                X = self._embed()
+            else:
+                X = self.X
+        elif self.is_root is False:
+            if self.reembed is True:
+                n_components = self._choose_ncomponents()
+                X = self._embed(n_components=n_components)
+            else:
+                X = self.X
+
+        if np.linalg.norm(X) < 1:
+            X = X / np.linalg.norm(X)
+
+        # self.X_ = X
+        results = self._cluster(
+            X,
+            cluster_method="GMM",
+            n_init=self.n_init,
+        )
+
+        self.results_ = results
+
+    def _to_adjacency(self):
+        if self.adj is None:
+            pass
+        # TODO: convert data matrix e.g. distance or similarity matrix
+        # to adjacency matrix
+
+    def _choose_ncomponents(self):
```

Comment:
this doesn't need to be a function, I don't think. might have to talk more about it to explain but i think this should happen automatically in emebed